### PR TITLE
Feat: QrCode Component

### PR DESCRIPTION
**Pull request by brendan-defi**
**What changed? Why?**
* Implemented components and utilities for creating beautiful QR codes

Default
image

Base
image

Cyberpunk
image

Hacker
image

**Notes to reviewers**

**How has it been tested?**
locally, and with test coverage

### CODE REVIEW

File: src/internal/components/QrCode/QrCodeSvg.tsx [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+  quietZone?: number;
+  quietZoneBorderRadius?: number;
+  ecl?: 'L' | 'M' | 'Q' | 'H';
+  gradientType?: 'radial' | 'linear';
+};
+
+export function QrCodeSvg({
+  value,
+  size = QR_CODE_SIZE,
+  backgroundColor = '#ffffff',
```
Comment (alessey) [inline src/internal/components/QrCode/QrCodeSvg.tsx:42]: should these default to theme colors to support dark mode?

File: src/internal/components/QrCode/gradientConstants.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+  },
+  black: {
+    startColor: '#a1a1aa',
+    endColor: '#27272a',
+  },
+  baseBlue: {
+    startColor: '#0052ff',
+    endColor: '#b2cbff',
+  },
+};
```
Comment (alessey) [inline src/internal/components/QrCode/gradientConstants.ts:42]: is it possible to make all of these pull from theme colors so when a user customizes a theme (or adds their own) the qr code adapts?

File: src/internal/components/QrCode/QrCodeSvg.tsx [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+                key={`${gradientColor}${offset}`}
+                offset={offset}
+                stopColor={gradientColor}
+                stopOpacity={1}
+              />
+            ))}
+          </radialGradient>
+        ) : (
+          <linearGradient
+            id="linearGrad"
```
Comment (alessey) [inline src/internal/components/QrCode/QrCodeSvg.tsx:143]: concat with [useId](https://react.dev/reference/react/useId) to ensure no conflicts

File: src/internal/components/QrCode/QrCodeSvg.tsx [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+  logoBackgroundColor?: string;
+  logoMargin?: number;
+  logoBorderRadius?: number;
+  quietZone?: number;
+  quietZoneBorderRadius?: number;
+  ecl?: 'L' | 'M' | 'Q' | 'H';
+  gradientType?: 'radial' | 'linear';
+};
+
+export function QrCodeSvg({
```
Comment (alessey) [inline src/internal/components/QrCode/QrCodeSvg.tsx:39]: may want to wrap in [memo](https://react.dev/reference/react/memo)

File: src/internal/components/QrCode/useCorners.tsx [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+      <g>
+        <rect
+          x={0}
+          y={0}
+          rx={9.5}
+          ry={9.5}
+          width={rectSize}
+          height={rectSize}
+          fill={fillColor}
+          id="Corner"
```
Comment (alessey) [inline src/internal/components/QrCode/useCorners.tsx:25]: ids should be unique

File: src/internal/components/QrCode/useDotsPath.test.ts [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    const { result: roundLogoResult } = renderHook(() =>
+      useDotsPath(withRoundLogo),
+    );
+    const { result: squareLogoResult } = renderHook(() =>
+      useDotsPath({ ...withRoundLogo, logoBorderRadius: 0 }),
+    );
+
+    expect(roundLogoResult.current).not.toBe(squareLogoResult.current);
+    // expect(roundLogoResult.current).not.toBe('');
+    // expect(squareLogoResult.current).not.toBe('');
```
Comment (alessey) [inline src/internal/components/QrCode/useDotsPath.test.ts:70]: nit: delete

File: src/internal/components/QrCode/useLogo.tsx [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+  logoSize,
+  logoBackgroundColor,
+  logoMargin,
+  logoBorderRadius,
+}: RenderLogoProps) {
+  const svgLogo = useMemo(() => {
+    if (React.isValidElement(logo)) {
+      logo = `data:image/svg+xml;charset=utf-8,${encodeURIComponent(
+        ReactDOMServer.renderToString(logo),
+      )}`;
```
Comment (alessey) [inline src/internal/components/QrCode/useLogo.tsx:31]: can't we just display the svg?

File: src/internal/components/QrCode/useMatrix.ts [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+
+export function useMatrix(
+  value: string,
+  errorCorrectionLevel: 'L' | 'M' | 'Q' | 'H',
+) {
+  if (!value) {
+    return [];
+  }
+
+  const matrix = useMemo(() => {
```
Comment (alessey) [inline src/internal/components/QrCode/useMatrix.ts:12]: can't conditionally call useMemo https://react.dev/reference/rules/rules-of-hooks

File: src/internal/components/QrCode/useMatrix.ts [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+    const arr = Array.prototype.slice.call(
+      QRCode.create(value, { errorCorrectionLevel }).modules.data,
+      0,
+    );
```
Comment (alessey) [inline src/internal/components/QrCode/useMatrix.ts:16]: why do you need to clone this?

**Review by dschlabach — APPROVED** [review]

File: src/internal/components/QrCode/QrCodeSvg.tsx [hunk 10]
```diff
@@ -0,0 +1,183 @@
+import { useMemo } from 'react';
```
Comment (cpcramer) [inline src/internal/components/QrCode/QrCodeSvg.tsx:1]: This should live in `src/internal/svg`

File: src/internal/components/QrCode/QrCodeSvg.tsx [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+  backgroundColor?: string;
+  logo?: React.ReactNode;
+  logoSize?: number;
+  logoBackgroundColor?: string;
+  logoMargin?: number;
+  logoBorderRadius?: number;
+  quietZone?: number;
+  quietZoneBorderRadius?: number;
+  ecl?: 'L' | 'M' | 'Q' | 'H';
+  gradientType?: 'radial' | 'linear';
```
Comment (cpcramer) [inline src/internal/components/QrCode/QrCodeSvg.tsx:36]: I know many of these are optional but curious if we even need all of these props - do we want this low level of customization for a QR code? I imagine that a few of these params will always stay the same and therefore we can just default in the func.

File: src/internal/components/QrCode/gradientConstants.ts (hunk 2 above)
Comment (cpcramer) [reply src/internal/components/QrCode/gradientConstants.ts:42]: Agreed - we should avoid having to update this file when adding new themes. 

Has the QR component been tested with custom themes? Curious if it's updating as expected.

File: src/internal/components/QrCode/QrCodeSvg.tsx (hunk 1 above)
Comment (brendan-defi) [reply src/internal/components/QrCode/QrCodeSvg.tsx:42]: eventually yes. unfortunately, enough of the code here assumes lightmode that getting to dark mode compatible was out of scope. this is what it currently looks like if we try to use the darkmode backgrounds:

image

dark-mode compatibility is going to be a fast-follow

**Review by alessey — APPROVED** [review]
approved with themes/dark mode as fast follows

**Pull request merged**